### PR TITLE
Add keyRotation spec to the StorageCluster

### PR DESCRIPTION
**Pull request by SanjalKatiyar**
https://issues.redhat.com/browse/RHSTOR-3891

Older PR: https://github.com/red-hat-storage/ocs-operator/pull/1926

### CODE REVIEW

File: controllers/storagecluster/noobaa_system_reconciler.go [hunk 1]
```diff
@@ -240,6 +241,9 @@ func (r *StorageClusterReconciler) setNooBaaDesiredState(nb *nbv1.NooBaa, sc *oc
 		}
 	}
 
+	isEnabled, rotationSchedule := util.GetKeyRotationSpec(sc)
+	nb.Spec.Security.KeyManagementService.EnableKeyRotation = isEnabled
+	nb.Spec.Security.KeyManagementService.Schedule = rotationSchedule
```
Comment (SanjalKatiyar) [inline controllers/storagecluster/noobaa_system_reconciler.go:246]: looks like NooBaa is always looking for the spec under KMS: https://github.com/noobaa/noobaa-operator/blob/master/pkg/apis/noobaa/v1alpha1/noobaa_types.go#L254

**Comment by SanjalKatiyar** [comment]
@iamniting @Rakshith-R ptal.

File: controllers/util/util.go [hunk 2]
```diff
@@ -22,3 +26,16 @@ func DetectDuplicateInStringSlice(slice []string) bool {
...
+		schedule := sc.Spec.Encryption.KeyRotation.Schedule
+		if schedule == "" {
+			// default schedule
+			schedule = "@Weekly"
+		}
+		// use key-rotation by default if cluster-wide encryption is opted without KMS
+		return true, schedule
+	}
+	return sc.Spec.Encryption.KeyRotation.Enabled, ""
+}
```
Comment (SanjalKatiyar) [inline controllers/util/util.go:41]: when we support KMS, we can update this utility function...

File: api/v1/storagecluster_types.go [hunk 3]
```diff
@@ -447,6 +447,19 @@ type EncryptionSpec struct {
...
+}
+
+// KeyRotationSpec represents the settings for Key Rotation.
+type KeyRotationSpec struct {
+	// Enabled represents whether the key rotation is enabled.
+	// +optional
+	Enabled bool `json:"enabled,omitempty"`
+	// Schedule represents the cron schedule for key rotation.
+	// +optional
+	Schedule string `json:"schedule,omitempty"`
```
Comment (Rakshith-R) [inline api/v1/storagecluster_types.go:462]: `omtiempty` is only helpful when the object is a pointer. We can remove it.

Let's add default to Schedule and keep it as @weekly using kubebuilder annotation too.

**Review by Rakshith-R — COMMENTED** [review]
just a small nit 
The change looks good to me, Thanks !

cc @baum @travisn

File: api/v1/storagecluster_types.go (hunk 3 above)
Comment (SanjalKatiyar) [reply api/v1/storagecluster_types.go:462]: I did not add default values on purpose, does not make sense to add it even if cluster-wide encryption is not opted by the user...

Comment (SanjalKatiyar) [reply api/v1/storagecluster_types.go:462]: instead let `GetKeyRotationSpec` handle it at runtime...

Comment (SanjalKatiyar) [reply api/v1/storagecluster_types.go:462]: let me know if you still want to add defaults using markers ?

**Review by iamniting — CHANGES_REQUESTED** [review]

File: api/v1/storagecluster_types.go [hunk 4]
```diff
@@ -447,6 +447,19 @@ type EncryptionSpec struct {
...
+	// KeyRotation defines options for Key Rotation.
+	// +optional
+	KeyRotation KeyRotationSpec `json:"keyRotation,omitempty"`
+}
+
+// KeyRotationSpec represents the settings for Key Rotation.
+type KeyRotationSpec struct {
+	// Enabled represents whether the key rotation is enabled.
+	// +optional
+	Enabled bool `json:"enabled,omitempty"`
```
[the remark this answers is not in the document]
Comment (SanjalKatiyar) [reply api/v1/storagecluster_types.go:459]: copied it from rook: https://github.com/rook/rook/blob/master/pkg/apis/ceph.rook.io/v1/types.go#L319 :/
will update here though !

File: api/v1/storagecluster_types.go (hunk 3 above)
Comment (Rakshith-R) [reply api/v1/storagecluster_types.go]: > let me know if you still want to add defaults using markers ?

Yes, lets use it so its explicitly visible in the CR.
( even if it is not in use )

Comment (SanjalKatiyar) [reply api/v1/storagecluster_types.go]: updated...
I did not remove `omtiempty` though, it is harmless and is used for almost every field in the StorageCluster struct...

**Review by iamniting — APPROVED** [review]

**Pull request merged**